ludwig/features/image_feature.py

```diff
@@ -166,110 +178,123 @@ def add_feature_data(
             user_specified_num_channels = True
             num_channels = preprocessing_parameters[NUM_CHANNELS]
+        else:
+            user_specified_num_channels = False
+            num_channels = first_img_num_channels
 
         assert isinstance(num_channels, int), ValueError(
-            'Number of image channels needs to be an integer')
+            'Number of image channels needs to be an integer'
+        )
+
+        return (should_resize,
+                width,
+                height,
+                num_channels,
+                user_specified_num_channels,
+                first_image)
+
+    @staticmethod
+    def add_feature_data(
+            feature,
+            dataset_df,
+            data,
+            metadata,
+            preprocessing_parameters
+    ):
+        set_default_value(
+            feature['preprocessing'],
+            'in_memory',
+            preprocessing_parameters['in_memory']
+        )
+        set_default_value(
+            feature['preprocessing'],
+            'num_processes',
+            preprocessing_parameters['num_processes']
+        )
+        csv_path = None
+        if hasattr(dataset_df, 'csv'):
+            csv_path = os.path.dirname(os.path.abspath(dataset_df.csv))
+
+        num_images = len(dataset_df)
+        if num_images == 0:
+            raise ValueError('There are no images in the dataset provided.')
+
+        first_image_path = dataset_df[feature['name']][0]
+        if csv_path is None and not os.path.isabs(first_image_path):
+            raise ValueError('Image file paths must be absolute')
+
+        first_image_path = get_abs_path(csv_path, first_image_path)
+
+        (should_resize,
+         width,
+         height,
+         num_channels,
+         user_specified_num_channels,
+         first_image) = \
+            ImageBaseFeature._finalize_preprocessing_parameters(
+                preprocessing_parameters, first_image_path
+            )
 
         metadata[feature['name']]['preprocessing']['height'] = height
         metadata[feature['name']]['preprocessing']['width'] = width
         metadata[feature['name']]['preprocessing'][
             'num_channels'] = num_channels
 
+        read_image_and_resize = partial(
+            ImageBaseFeature._read_image_and_resize,
+            img_width=width,
+            img_height=height,
+            should_resize=should_resize,
+            num_channels=num_channels,
+            resize_method=preprocessing_parameters['resize_method'],
+            user_specified_num_channels=user_specified_num_channels
+        )
+        all_file_paths = [get_abs_path(csv_path, file_path)
+                          for file_path in dataset_df[feature['name']]]
+
         if feature['preprocessing']['in_memory']:
+            # Number of processes to run in parallel for preprocessing
+            num_processes = feature['preprocessing']['num_processes']
```

Comment:
Just curious, can we do parallel processing when images are not in memory?

Reply:
HDF5 supports concurrent reads. But anyway, this whole thing will change if we use petastorm for preprocessing.

Reply:
Yeah, it is possible. But it's not as straightforward. Let me mark a todo there